modules/tensor_mechanics/src/materials/HillElastoPlasticityStressUpdate.C

```diff
@@ -256,87 +275,114 @@ ADHillElastoPlasticityStressUpdate::computeOmega(const ADReal & delta_gamma,
 
-  ADReal omega = computeOmega(delta_gamma, stress_tilde);
+  GenericReal<is_ad> omega = computeOmega(delta_gamma, stress_tilde);
   _hardening_slope = computeHardeningDerivative();
 
   // Hardening variable is \alpha isotropic hardening for now.
   _hardening_variable[_qp] = computeHardeningValue(delta_gamma, omega);
-  ADReal s_y = _hardening_slope * _hardening_variable[_qp] + _yield_stress;
 
-  ADReal residual = 0.0;
+  // A small value of 1.0e-30 is added to the hardening variable to avoid numerical issues
+  // related to hardening variable becoming negative early in the iteration leading to non-
+  // convergence. Note that std::pow(x,n) requires x to be positive if n is less than 1.
+
+  GenericReal<is_ad> s_y =
+      _hardening_constant * std::pow(_hardening_variable[_qp] + 1.0e-30, _hardening_exponent) +
+      _yield_stress;
+
+  GenericReal<is_ad> residual = 0.0;
   residual = s_y / omega - 1.0;
 
   return residual;
 }
 
-ADReal
-ADHillElastoPlasticityStressUpdate::computeDerivative(const ADDenseVector & /*stress_dev*/,
-                                                      const ADDenseVector & stress_sigma,
-                                                      const ADReal & delta_gamma)
+template <bool is_ad>
+GenericReal<is_ad>
+HillElastoPlasticityStressUpdateTempl<is_ad>::computeDerivative(
+    const GenericDenseVector<is_ad> & /*stress_dev*/,
+    const GenericDenseVector<is_ad> & stress_sigma,
+    const GenericReal<is_ad> & delta_gamma)
 {
   // If in elastic regime, return unit derivative
   if (_yield_condition <= 0.0)
     return 1.0;
 
-  ADReal omega = computeOmega(delta_gamma, stress_sigma);
+  GenericReal<is_ad> omega = computeOmega(delta_gamma, stress_sigma);
   _hardening_slope = computeHardeningDerivative();
 
-  ADReal sy = _hardening_slope * computeHardeningValue(delta_gamma, omega) + _yield_stress;
-  ADReal sy_alpha = _hardening_slope;
+  GenericReal<is_ad> hardeningVariable = computeHardeningValue(delta_gamma, omega);
+  GenericReal<is_ad> sy =
+      _hardening_constant * std::pow(hardeningVariable + 1.0e-30, _hardening_exponent) +
+      _yield_stress;
+  GenericReal<is_ad> sy_alpha = _hardening_constant *
+                                std::pow(hardeningVariable + 1.0e-30, _hardening_exponent - 1) *
+                                _hardening_exponent;
 
-  ADReal omega_gamma;
-  ADReal sy_gamma;
+  GenericReal<is_ad> omega_gamma;
+  GenericReal<is_ad> sy_gamma;
 
   computeDeltaDerivatives(delta_gamma, stress_sigma, sy_alpha, omega, omega_gamma, sy_gamma);
-  ADReal residual_derivative = 1 / omega * (sy_gamma - 1 / omega * omega_gamma * sy);
+  GenericReal<is_ad> residual_derivative = 1 / omega * (sy_gamma - 1 / omega * omega_gamma * sy);
 
   return residual_derivative;
 }
 
+template <bool is_ad>
+GenericReal<is_ad>
+HillElastoPlasticityStressUpdateTempl<is_ad>::computeStressDerivative(
+    const Real /*effective_trial_stress*/, const Real /*scalar*/)
+{
+  return 1.0;
```

Comment:
Do we have an analogous method for the `HillPlasticityStressUpdate` class?

Reply:
This method will be used for calculating the tangent operator; since I haven't yet added the calculation for tangent operator, this method is not doing anything except make the coverage look bad. So, I'm removing it for now.